src/hotspot/share/runtime/thread.cpp

```diff
@@ -1585,157 +1581,109 @@ void JavaThread::remove_monitor_chunk(MonitorChunk* chunk) {
-    if ((_suspend_flags & _async_delivery_disabled) != 0) {
-      log_info(exceptions)("Async exception delivery is disabled");
+  if (has_last_Java_frame()) {
+    frame f = last_frame();
+    if (f.is_runtime_frame()) {
+      // If the topmost frame is a runtime stub, then we are calling into
+      // OptoRuntime from compiled code. Some runtime stubs (new, monitor_exit..)
+      // must deoptimize the caller before continuing, as the compiled exception
+      // handler table may not be valid.
+      RegisterMap reg_map(this, false);
+      frame compiled_frame = f.sender(&reg_map);
+      if (!StressCompiledExceptionHandlers && compiled_frame.can_be_deoptimized()) {
+        Deoptimization::deoptimize(this, compiled_frame);
+      }
     }
-    return;
   }
 
-  if (_pending_async_exception != NULL) {
-    // Only overwrite an already pending exception if it is not a threadDeath.
-    if (!has_pending_exception() || !pending_exception()->is_a(vmClasses::ThreadDeath_klass())) {
+  // Only overwrite an already pending exception if it is not a threadDeath.
+  if (!has_pending_exception() || !pending_exception()->is_a(vmClasses::ThreadDeath_klass())) {
 
-      // We cannot call Exceptions::_throw(...) here because we cannot block
-      set_pending_exception(_pending_async_exception, __FILE__, __LINE__);
+    // We cannot call Exceptions::_throw(...) here because we cannot block
+    set_pending_exception(java_throwable, __FILE__, __LINE__);
 
-      LogTarget(Info, exceptions) lt;
-      if (lt.is_enabled()) {
-        ResourceMark rm;
-        LogStream ls(lt);
-        ls.print("Async. exception installed at runtime exit (" INTPTR_FORMAT ")", p2i(this));
-          if (has_last_Java_frame()) {
-            frame f = last_frame();
-           ls.print(" (pc: " INTPTR_FORMAT " sp: " INTPTR_FORMAT " )", p2i(f.pc()), p2i(f.sp()));
-          }
-        ls.print_cr(" of type: %s", _pending_async_exception->klass()->external_name());
+    LogTarget(Info, exceptions) lt;
+    if (lt.is_enabled()) {
+      ResourceMark rm;
+      LogStream ls(lt);
+      ls.print("Async. exception installed at runtime exit (" INTPTR_FORMAT ")", p2i(this));
+      if (has_last_Java_frame()) {
+        frame f = last_frame();
+        ls.print(" (pc: " INTPTR_FORMAT " sp: " INTPTR_FORMAT " )", p2i(f.pc()), p2i(f.sp()));
       }
+      ls.print_cr(" of type: %s", java_throwable->klass()->external_name());
     }
-    // Always null out the _pending_async_exception oop here since the async condition was
-    // already cleared above and thus considered handled.
-    _pending_async_exception = NULL;
-  } else {
-    assert(_is_unsafe_access_error, "must be");
-    DEBUG_ONLY(_is_unsafe_access_error = false);
-
-    // We may be at method entry which requires we save the do-not-unlock flag.
-    UnlockFlagSaver fs(this);
-    Exceptions::throw_unsafe_access_internal_error(this, __FILE__, __LINE__, "a fault occurred in an unsafe memory access operation");
```

Comment:
This seems to be dead code in Exceptions now.

Also the UnlockFlagSaver that was moved into thread.hpp could now be moved back to being a local class in interpreterRuntime.cpp.

Reply:
Removed method from Exceptions::throw_unsafe_access_internal_error(). I restored the UnlockFlagSaver though which I shouldn't have removed. I looked at the crash from 8246381 where this line was introduced and thought this cannot happen anymore after 8274379, since we will always check async exceptions on return from native to Java. But there is still the case where the unsafe access could happen from compiled Java code rather than from vm/native. If after that unsafe access we make a call to some synchronized method, interpreted version, then I think we could have the same issue.